### PR TITLE
[libos] Add socket options

### PR DESCRIPTION
**Pull request by iyzhang**
This PR adds support for socket options to our stack. So far only the Rust API is implemented but support for SO_LINGER has been implemented for all libOSes (except for the inetstack which still needs its timed wait turned on).

### CODE REVIEW

File: src/rust/inetstack/mod.rs [hunk 1]
```diff
@@ -282,6 +283,27 @@ impl<N: NetworkRuntime> NetworkTransport for SharedInetStack<N> {
         }
     }
 
+    /// Set an SO_* option on the socket.
+    fn set_socket_option(&mut self, sd: &mut Self::SocketDescriptor, option: SocketOption) -> Result<(), Fail> {
+        match sd {
+            Socket::Tcp(socket) => self.ipv4.tcp.set_socket_option(socket, option),
+            Socket::Udp(_) => Err(Fail::new(libc::ENOTSUP, "No socket options for UDP are supported")),
```
Comment (ppenna) [inline src/rust/inetstack/mod.rs:290]: Could you change this to log a `error()` message before failling?

File: src/rust/inetstack/mod.rs [hunk 2]
```diff
@@ -282,6 +283,27 @@ impl<N: NetworkRuntime> NetworkTransport for SharedInetStack<N> {
...
+    /// Gets an SO_* option on the socket. The option should be passed in as [option] and the value is returned in
+    /// [option].
+    fn get_socket_option(
+        &mut self,
+        sd: &mut Self::SocketDescriptor,
+        option: SocketOption,
+    ) -> Result<SocketOption, Fail> {
+        match sd {
+            Socket::Tcp(socket) => self.ipv4.tcp.get_socket_option(socket, option),
+            Socket::Udp(_) => Err(Fail::new(libc::ENOTSUP, "No socket options for UDP are supported")),
```
Comment (ppenna) [inline src/rust/inetstack/mod.rs:303]: Could you change this to log a `error()` message before failling?

File: src/rust/demikernel/libos/mod.rs [hunk 3]
```diff
@@ -246,6 +247,54 @@ impl LibOS {
...
+            match self {
+                #[cfg(any(
+                    feature = "catnap-libos",
+                    feature = "catnip-libos",
+                    feature = "catpowder-libos",
+                    feature = "catloop-libos"
+                ))]
+                LibOS::NetworkLibOS(libos) => libos.set_socket_option(sockqd, option),
+                #[cfg(feature = "catmem-libos")]
+                LibOS::MemoryLibOS(_) => Err(Fail::new(
```
Comment (ppenna) [inline src/rust/demikernel/libos/mod.rs:262]: Could you change this to log a `error()` message before failling?

File: src/rust/demikernel/libos/mod.rs [hunk 4]
```diff
@@ -246,6 +247,54 @@ impl LibOS {
...
+            match self {
+                #[cfg(any(
+                    feature = "catnap-libos",
+                    feature = "catnip-libos",
+                    feature = "catpowder-libos",
+                    feature = "catloop-libos"
+                ))]
+                LibOS::NetworkLibOS(libos) => libos.get_socket_option(sockqd, option),
+                #[cfg(feature = "catmem-libos")]
+                LibOS::MemoryLibOS(_) => Err(Fail::new(
```
Comment (ppenna) [inline src/rust/demikernel/libos/mod.rs:286]: Could you change this to log a `error()` message before failling?

File: src/rust/catnap/linux/transport.rs [hunk 5]
```diff
@@ -336,6 +339,38 @@ impl NetworkTransport for SharedCatnapTransport {
...
+        &mut self,
+        sd: &mut Self::SocketDescriptor,
+        option: SocketOption,
+    ) -> Result<SocketOption, Fail> {
+        trace!("Set socket option to {:?}", option);
+        let socket: &mut Socket = self.socket_from_sd(sd);
+        match option {
+            SocketOption::SO_LINGER(_) => match socket.linger() {
+                Ok(linger) => Ok(SocketOption::SO_LINGER(linger)),
+                Err(e) => Err(Fail::new(get_libc_err(e), "set_socket_option(SO_LINGER): ")),
```
Comment (ppenna) [inline src/rust/catnap/linux/transport.rs:369]: Could you change this to log a `error()` message before failling?

File: src/rust/runtime/network/socket/option.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+pub struct TcpSocketOptions {
+    linger: Option<Duration>,
+}
+
+impl TcpSocketOptions {
+    pub fn get_linger(&self) -> Option<Duration> {
+        self.linger
+    }
+
+    pub fn set_linger(&mut self, linger: Option<Duration>) {
```
Comment (ppenna) [inline src/rust/runtime/network/socket/option.rs:34]: No check is performed if passed linger is `None`. What is the expected behavior for this case? How does that differentiate from setting linger to `0`?

File: examples/tcp-echo/server.rs [hunk 7]
```diff
@@ -68,6 +74,8 @@ impl TcpEchoServer {
     pub fn new(mut libos: LibOS, local: SocketAddr) -> Result<Self> {
         // Create a TCP socket.
         let sockqd: QDesc = libos.socket(AF_INET, SOCK_STREAM, 0)?;
+        
+        libos.set_socket_option(sockqd, SocketOption::SO_LINGER(Some(DEFAULT_LINGER)))?;
```
Comment (ppenna) [inline examples/tcp-echo/server.rs:78]: Please remove trailling spaces.

File: examples/tcp-wait/server.rs [hunk 8]
```diff
@@ -23,6 +23,9 @@ use ::std::{
     },
     net::SocketAddr,
 };
+use demikernel::runtime::network::socket::option::SocketOption;
```
Comment (ppenna) [inline examples/tcp-wait/server.rs:26]: If `SocketOption` will be exposed to user-applications, could you re-export it at the `libs.rs` level?

**Review by ppenna — COMMENTED** [review]
Could you change the `bindings.rs` as well?

File: src/rust/runtime/network/socket/option.rs (hunk 6 above)
Comment (iyzhang) [reply src/rust/runtime/network/socket/option.rs:34]: If linger is None, then the default is to wait for 2 * MSL, which I implement later in timed wait.

File: src/rust/demikernel/bindings.rs [hunk 9]
```diff
@@ -793,15 +789,63 @@ pub extern "C" fn demi_setsockopt(
...
+            if optlen as usize != mem::size_of::<Linger>() {
+                warn!("demi_setsockopt() linger len is incorrect");
+                return libc::EINVAL;
+            }
+
+            let linger: Linger = unsafe { *(optval as *const Linger) };
+            match linger.l_onoff {
+                0 => SocketOption::SO_LINGER(None),
+                _ => SocketOption::SO_LINGER(Some(Duration::from_secs(linger.l_linger as u64))),
+            }
```
Comment (ppenna) [inline src/rust/demikernel/bindings.rs:818]: Going back to my previous comment. How does one set linger to 0, if `Non` means default value?

Comment (iyzhang) [reply src/rust/demikernel/bindings.rs:818]: you set it to Some(Duration::ZERO)

**Review by ppenna — APPROVED** [review]
Thanks for working on this. Could you please squash intermediate commits before merging?

**Pull request merged**